### PR TITLE
[quant] QNNPACK deconv kernel and tests

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37497 [quant] QNNPACK factoring out common structs from conv/deconv
* **#36790 [quant] QNNPACK deconv kernel and tests**
* #37405 [quant] QNNPACK deconvolution packing

Differential Revision: [D21110111](https://our.internmc.facebook.com/intern/diff/D21110111)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:156]: @kimishpatel does the float requantization PR apply to the deconv operation too?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:181]: Can we use nullptr instead of NULL in this file?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:319]: nit: message

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:67]: can we move this struct to a common header now that's used in 2 places?

**Review by supriyar — CHANGES_REQUESTED** [review]
Left some comments. Can you add test for this API in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/test/deconvolution-operator-tester.h?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:156]: Yes it does. And it is covered under conv operator. Please see this PR: https://github.com/pytorch/pytorch/pull/35856. I suggest we wait for it to land? Else I will have to make fixes here.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:96]: BTW, why are we doing these checks here? Can this not to be done in deconv_param_t constructor. Conv op does that so we should probably maintain the convention. Even better if we can just use conv_params_t if possible.

Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:96]: Yeah I think @z-a-f  mentioned this as a todo in the previous PR. I think we can merge them

Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:96]: This check is done in the params. I can remove it from here

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:67]: I would love to. The problem is that the qnnpack library is compiled using C11, which doesn't support operator overloading. The `cc` files are compiled using CXX14, which does support it. That's why it is OK to keep the deleter here, but not in the common. I will look into it, because I agree that factoring it out is a good idea -- just need to find a way

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:156]: Adding a TODO in here in case this lands before #35856

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:67]: I will send a separate PR for this -- it looks like it is a little more involved.

Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:67]: Please add a TODO or file an issue to track it

**Comment by supriyar** [comment]
@z-a-f can we add tests in QNNPACK to test the new APIs added? Otherwise the PR looks good to me.

**Comment by kimishpatel** [comment]
Can we add test for both inside QNNPACK that can compile for armv7/v8 and python test much like other ones you have?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:30]: There is already a compute_q8conv in conv-run.cc. Lets use that if possible, please?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:67]: This might sound like a rant, but let's prioritize deduplication of code. Even the current qnnpack state seems little unmaintainable.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:124]: #35856 has landed.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:73]: I think this and qnnpackConv probably can merge as well with appropriate refactoring, or atleast overlap can be minimized. We can put this under a common umbrella of conv/deconv merge and have a PR to fix that.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:247]: No need for switch here. Only single case.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Left some comments with requested changes.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:124]: Can you elaborate what changes the PR will require?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:73]: These two cannot be merged -- they are two different kernels. I can refactor stuff, but refactoring will be a part of future PRs.

Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:73]: I agree with the overlap -- but I don't think this can be merged with the conv -- these are two different ops, and should live in their own classes.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:67]: > Please add a TODO or file an issue to track it

Wil  send a separate PR -- we need to discuss what parts of qnnpack should be made available to PT

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:124]: I was gonna say like this: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc#L308. But it seems that I missed removing this check since it is not longer a requirement. @supriyar, did this not trigger on the model we ran the PR for requant changes? Seems strange.
@z-a-f I think you can remove this check.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:73]: I need to think through this. If we have a way to share as much code as possible that would be good. On mobile side this is also an app size issue. From clean code perspective I understand. We shouldn't merge things and make a bloated structures that house unrelated components. For now this is fine.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:124]: >> @supriyar, did this not trigger on the model we ran the PR for requant changes? Seems strange.

I had made the changes locally when I ran the model so wasn't an issue.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:30]: factored out in #37497

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:124]: @kimishpatel I am confused, we don't need this condition anymore? Can I just remove the if statement defined in line 156 completely?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:67]: Moved to the refactoring PR

**Comment by z-a-f** [comment]
> @z-a-f can we add tests in QNNPACK to test the new APIs added? Otherwise the PR looks good to me.

I am ironing out some stuff related to the testing. I wanted to add the test in this PR (for future bisecting purposes), but because the prepacking is outside of this PR, I might send the tests separately. I will try to rearrange so that the prepack would be before this -- lemme try

**Comment by z-a-f** [comment]
> Can we add test for both inside QNNPACK that can compile for armv7/v8 and python test much like other ones you have?

Python tests cannot use this kernels directly until the python wrappers are written. Those will come later when the python wrappers are written

**Comment by z-a-f** [comment]
> @z-a-f can we add tests in QNNPACK to test the new APIs added? Otherwise the PR looks good to me.

Done

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:123]: Yes, we dont need this condition anymore. Please remove it. I need to remove similar ones from conv-run and fc-run.

Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:123]: Will do before landing

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/deconvolution-operator-tester.h
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/test/deconvolution-operator-tester.h:564]: I think you also need to call the tests with Runtime mode

File: aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt:433]: Why do we need fp16 library for this?

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution.cc
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution.cc:15]: So we're enabling both static and runtime mode for all conv and deconv tests. Do they all pass? I recall you mentioning some of them failed

**Review by supriyar — APPROVED** [review]
Looks good, left some comments in the review.

**Review by raghuramank100 — APPROVED** [review]

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution.cc
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution.cc:15]: The same that fail in the regular convolution fail in the transpose as well. I think this is not specific to convolution. Currently, I am enabling the same tests that are enabled from the conv

File: aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt
Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt:433]: FP16 is required for the QNNPACK requantization

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution.cc
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/test/convolution.cc:15]: Are you sure? Check this on master? On master I think all convolution tests are passing. If it is not failing on master it should not be failing here.

**Pull request closed without merging**
